dronegen/push.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import "fmt"
+
+// pushCheckoutCommands builds a list of commands for Drone to check out a git commit on a push build
+func pushCheckoutCommands(fips bool) []string {
+	commands := []string{
+		`mkdir -p /go/src/github.com/gravitational/teleport /go/cache`,
+		`cd /go/src/github.com/gravitational/teleport`,
+		`git init && git remote add origin ${DRONE_REMOTE_URL}`,
+		`git fetch origin`,
+		`git checkout -qf ${DRONE_COMMIT_SHA}`,
+		// this is allowed to fail because pre-4.3 Teleport versions don't use the webassets submodule
+		`git submodule update --init webassets || true`,
+		`mkdir -m 0700 /root/.ssh && echo "$GITHUB_PRIVATE_KEY" > /root/.ssh/id_rsa && chmod 600 /root/.ssh/id_rsa`,
+		`ssh-keyscan -H github.com > /root/.ssh/known_hosts 2>/dev/null && chmod 600 /root/.ssh/known_hosts`,
+		`git submodule update --init e`,
+		// do a recursive submodule checkout to get both webassets and webassets/e
+		// this is allowed to fail because pre-4.3 Teleport versions don't use the webassets submodule
+		`git submodule update --init --recursive webassets || true`,
+		`rm -f /root/.ssh/id_rsa`,
+	}
+	if fips {
+		commands = append(commands, `if [[ "${DRONE_TAG}" != "" ]]; then echo "${DRONE_TAG##v}" > /go/.version.txt; else egrep ^VERSION Makefile | cut -d= -f2 > /go/.version.txt; fi; cat /go/.version.txt`)
+	}
+	return commands
+}
+
+// pushBuildCommands generates a list of commands for Drone to build an artifact as part of a push build
+func pushBuildCommands(b buildType) []string {
+	commands := []string{
+		`apk add --no-cache make`,
+		`chown -R $UID:$GID /go`,
+		`cd /go/src/github.com/gravitational/teleport`,
+	}
+	if b.fips {
+		commands = append(commands,
+			`export VERSION=$(cat /go/.version.txt)`,
+		)
+	}
+	commands = append(commands,
+		fmt.Sprintf(`make -C build.assets %s`, releaseMakefileTarget(b)),
+	)
+	return commands
+}
+
+// pushPipelines builds all applicable push pipeline combinations
+func pushPipelines() []pipeline {
+	var ps []pipeline
+	for _, arch := range []string{"amd64", "386", "arm", "arm64"} {
+		for _, fips := range []bool{false, true} {
+			if (arch == "386" || arch == "arm") && fips {
+				// FIPS mode not supported on i386/ARM
+				continue
+			}
+			ps = append(ps, pushPipeline(buildType{os: "linux", arch: arch, fips: fips}))
+		}
+	}
+	// Only amd64 Windows is supported for now.
+	ps = append(ps, pushPipeline(buildType{os: "windows", arch: "amd64"}))
+	return ps
+}
+
+// pushPipeline generates a push pipeline for a given combination of os/arch/FIPS
+func pushPipeline(b buildType) pipeline {
+	if b.os == "" {
+		panic("b.os must be set")
+	}
+	if b.arch == "" {
+		panic("b.arch must be set")
+	}
+
+	pipelineName := fmt.Sprintf("push-build-%s-%s", b.os, b.arch)
+	pushEnvironment := map[string]value{
```

Comment:
nit: these could also be factored out as UID/GID

Reply:
Yeah, once again I'm just trying to generate the exact same pipeline as we currently have to minimise the chance of things breaking.